### PR TITLE
Resolve missing OPENPYPE_MONGO in deadline global job preload 

### PR DESCRIPTION
**Pull request by mre7a**
## Brief description
In the GlobalJobPreLoad plugin, we propose to replace the SpawnProcess by a sub-process and to pass the environment variables in the parameters, since the SpawnProcess under Centos Linux does not pass the environment variables.

## Description
In the GlobalJobPreLoad plugin, the Deadline SpawnProcess is used to start the OpenPype process. The problem is that the SpawnProcess does not pass environment variables, including OPENPYPE_MONGO, to the process when it is under Centos7 linux, and the process gets stuck. We propose to replace it by a subprocess and to pass the variable in the parameters.

## Additional info
Please note that I have not tested it on Windows. 

## Testing notes:
1. Run a deadline Job, under Linux, Centos 7, Deadline 10.1
2. Check if OPENPYPE_MONGO is in the os environment. 
3. Check if the spawned process can see the variable.

### CODE REVIEW

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py [hunk 1]
```diff
@@ -190,17 +189,23 @@ def inject_openpype_environment(deadlinePlugin):
...
-            exe, args_str, os.path.dirname(exe)
+
+        current_env = {**os.environ}
+        current_env["OPENPYPE_MONGO"] = job.GetJobEnvironmentKeyValue(
+            'OPENPYPE_MONGO'
         )
-        ProcessUtils.WaitForExit(process, -1)
-        if process.ExitCode != 0:
+
+        process = subprocess.Popen([exe] + args, env=current_env)
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:205]: Will deadline monitor show output from the process when `subprocess.Popen` is used?

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py [hunk 2]
```diff
@@ -190,17 +189,23 @@ def inject_openpype_environment(deadlinePlugin):
             raise RuntimeError(msg)
 
         if not os.environ.get("OPENPYPE_MONGO"):
-            print(">>> Missing OPENPYPE_MONGO env var, process won't work")
+            if not job.GetJobEnvironmentKeyValue('OPENPYPE_MONGO'):
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:192]: Why is `if not os.environ.get("OPENPYPE_MONGO"):` still checked when it's 100% time overriden by `job.GetJobEnvironmentKeyValue('OPENPYPE_MONGO')`?

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py [hunk 3]
```diff
@@ -190,17 +189,23 @@ def inject_openpype_environment(deadlinePlugin):
...
 
         os.environ["AVALON_TIMEOUT"] = "5000"
 
         args_str = subprocess.list2cmdline(args)
         print(">>> Executing: {} {}".format(exe, args_str))
-        process = ProcessUtils.SpawnProcess(
-            exe, args_str, os.path.dirname(exe)
+
+        current_env = {**os.environ}
+        current_env["OPENPYPE_MONGO"] = job.GetJobEnvironmentKeyValue(
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:201]: You can just set `os.environ`. It's single time use process and you don't need to worry aby having any affect.
```suggestion
        os.environ["OPENPYPE_MONGO"] = job.GetJobEnvironmentKeyValue(
```

**Comment by iLLiCiTiT** [comment]
> since the SpawnProcess under Centos Linux does not pass the environment variables.

It does not pass environment variables to spawned process or it does not load environments from job? (that's 2 different issues)

**Comment by mre7a** [comment]
> It does not pass environment variables to spawned process or it does not load environments from job enviroments? (that's 2 different issues)
>

I'll try to describe the problem in detail. Originally, without my modifications, and with SpawnProcess, we have this error log on our job:
```
=======================================================
Error
=======================================================
Error: Global job preload script '/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalJobPreLoad.py': RuntimeError : Failed to run OpenPype process to extract environments. (FranticX.Scripting.PythonNetException)
  File "/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalJobPreLoad.py", line 434, in __main__
    inject_openpype_environment(deadlinePlugin)
  File "/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalJobPreLoad.py", line 368, in inject_openpype_environment
    "Failed to run OpenPype process to extract environments."
   at Python.Runtime.PyObject.Invoke(PyObject[] args)
   at Python.Runtime.PyObject.InvokeMethod(String name, PyObject[] args)
   at FranticX.Scripting.PythonNetScriptEngine.CallFunction(String moduleName, String functionName, Object[] args)
   at Deadline.Plugins.PluginWrapper.d()
   at Deadline.Plugins.PluginWrapper.StartJob(String& outMessage, AbortLevel& abortLevel)

=======================================================
Type
=======================================================
RenderPluginException

=======================================================
Stack Trace
=======================================================
   at Deadline.Plugins.SandboxedPlugin.d(DeadlineMessage bgm, CancellationToken bgn)
   at Deadline.Plugins.SandboxedPlugin.StartJob(Job job, CancellationToken cancellationToken)
   at Deadline.Slaves.SlaveRenderThread.c(TaskLogWriter ajt, CancellationToken aju)

=======================================================
Log
=======================================================
2023-02-23 11:36:51:  0: Loading Job's Plugin timeout is Disabled
2023-02-23 11:36:51:  0: SandboxedPlugin: Render Job As User enabled, running as user 'xxxxxxx', which is defined in the User Settings for this job's user
2023-02-23 11:36:56:  0: Executing plugin command of type 'Initialize Plugin'
2023-02-23 11:36:56:  0: INFO: Executing plugin script '/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/Nuke.py'
2023-02-23 11:36:56:  0: INFO: Plugin execution sandbox using Python version 3
2023-02-23 11:36:56:  0: INFO: About: Nuke Plugin for Deadline
2023-02-23 11:36:56:  0: INFO: The job's environment will be merged with the current environment before rendering
2023-02-23 11:36:56:  0: Done executing plugin command of type 'Initialize Plugin'
2023-02-23 11:36:56:  0: Start Job timeout is disabled.
2023-02-23 11:36:56:  0: Task timeout is disabled.
2023-02-23 11:36:56:  0: Loaded job: xxxx_001_0010_workfileReference_v028.nk - renderReferenceMain (63f740fb9caa9f05f0595caa)
2023-02-23 11:36:56:  0: Executing plugin command of type 'Start Job'
2023-02-23 11:36:56:  0: DEBUG: S3BackedCache Client is not installed.
2023-02-23 11:36:56:  0: INFO: Executing global asset transfer preload script '/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalAssetTransferPreLoad.py'
2023-02-23 11:36:57:  0: INFO: Looking for legacy (pre-10.0.26) AWS Portal File Transfer...
2023-02-23 11:36:57:  0: INFO: Looking for legacy (pre-10.0.26) File Transfer controller in /opt/Thinkbox/S3BackedCache/bin/task.py...
2023-02-23 11:36:57:  0: INFO: Could not find legacy (pre-10.0.26) AWS Portal File Transfer.
2023-02-23 11:36:57:  0: INFO: Legacy (pre-10.0.26) AWS Portal File Transfer is not installed on the system.
2023-02-23 11:36:57:  0: INFO: Executing global job preload script '/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalJobPreLoad.py'
2023-02-23 11:36:57:  0: PYTHON: *** GlobalJobPreload start ...
2023-02-23 11:36:57:  0: PYTHON: >>> Getting job ...
2023-02-23 11:36:57:  0: PYTHON: --- Job type - render 1
2023-02-23 11:36:57:  0: PYTHON: --- Job type - publish 0
2023-02-23 11:36:57:  0: PYTHON: --- Job type - remote 0
2023-02-23 11:36:57:  0: PYTHON: >>> Injecting OpenPype environments ...
2023-02-23 11:36:57:  0: PYTHON: >>> Scanning for compatible requested version 3.14.7
2023-02-23 11:36:57:  0: PYTHON: >>> Getting OpenPype executable ...
2023-02-23 11:36:57:  0: PYTHON: --- Looking for OpenPype at: xxxx/apps/openpype/latest
2023-02-23 11:36:57:  0: PYTHON:   - found: <OpenPypeVersion 3.14.7-quad.10> - xxxx/apps/openpype/latest/linux
2023-02-23 11:36:57:  0: PYTHON:    ! path is not a build: xxxx/apps/openpype/latest/windows
2023-02-23 11:36:57:  0: PYTHON: Looking for OpenPype at: xxxx/apps/openpype/latest/linux
2023-02-23 11:36:57:  0: PYTHON:   - found: <OpenPypeVersion 3.14.7-quad.10> - xxxx/apps/openpype/latest/linux
2023-02-23 11:36:57:  0: PYTHON: *** Latest compatible version found is xxxx/apps/openpype/latest/linux in <OpenPypeVersion 3.14.7-quad.10>
2023-02-23 11:36:57:  0: PYTHON: --- OpenPype executable: xxxx/apps/openpype/latest/linux/openpype_console
2023-02-23 11:36:57:  0: PYTHON: >>> Temporary path: /tmp/20230223103657303122_ff3c4e40-b365-11ed-a68b-08c0eb739cb6.json
2023-02-23 11:36:57:  0: PYTHON: >>> Missing OPENPYPE_MONGO env var, process won't work
2023-02-23 11:36:57:  0: PYTHON: >>> Executing: xxxx/apps/openpype/latest/linux/openpype_console --headless extractenvironments /tmp/20230223103657303122_ff3c4e40-b365-11ed-a68b-08c0eb739cb6.json --project TEST_OP_PUB --asset 001_0010 --task Reference --app nuke/14-0 --envgroup farm
2023-02-23 11:37:00:  0: PYTHON: Traceback (most recent call last):
2023-02-23 11:37:00:  0: PYTHON:   File "/tmp/deadline/nodexxx/plugins/63f740fb9caa9f05f0595caa/GlobalJobPreLoad.py", line 368, in inject_openpype_environment
2023-02-23 11:37:00:  0: PYTHON:     "Failed to run OpenPype process to extract environments."
2023-02-23 11:37:00:  0: PYTHON: RuntimeError: Failed to run OpenPype process to extract environments.
2023-02-23 11:37:00:  0: PYTHON: !!! Injection failed.
2023-02-23 11:37:00:  0: 'xxxxxxx/Thinkbox/Deadline10/pythonAPIs/8V8myYiuCTeqVADSSX273A==' already exists. Skipping extraction of PythonSync.
2023-02-23 11:37:02:  0: Done executing plugin command of type 'Start Job'


=======================================================
Worker Information
=======================================================
Version: v10.1.23.6 Release (773a6289d)
Operating System: Linux
Machine User: root
CPU Architecture: x86_64
```
And I can see in job properties, that I have OPENPYPE_MONGO variable. Same error (without Missing OPENPYPE_MONGO part) if I add it manually to `os.environ`. 

I did not find a way to add environment to SpawnProcess, and it works fine just after replacing it with a subprocess.Popen. 
I add the log with subprocess below. 

```
2023-02-23 16:16:11:  0: PYTHON: *** GlobalJobPreload start ...
2023-02-23 16:16:11:  0: PYTHON: >>> Getting job ...
2023-02-23 16:16:11:  0: PYTHON: --- Job type - render 1
2023-02-23 16:16:11:  0: PYTHON: --- Job type - publish 0
2023-02-23 16:16:11:  0: PYTHON: --- Job type - remote 0
2023-02-23 16:16:11:  0: PYTHON: >>> Injecting OpenPype environments ...
2023-02-23 16:16:11:  0: PYTHON: >>> Scanning for compatible requested version 3.14.7
2023-02-23 16:16:11:  0: PYTHON: >>> Getting OpenPype executable ...
2023-02-23 16:16:11:  0: PYTHON: --- Looking for OpenPype at: xxxxx/apps/openpype/latest
2023-02-23 16:16:11:  0: PYTHON:   - found: <OpenPypeVersion 3.14.7-quad.10> - xxxxx/apps/openpype/latest/linux
2023-02-23 16:16:11:  0: PYTHON:    ! path is not a build: xxxxx/apps/openpype/latest/windows
2023-02-23 16:16:11:  0: PYTHON: Looking for OpenPype at: xxxxx/apps/openpype/latest/linux
2023-02-23 16:16:11:  0: PYTHON:   - found: <OpenPypeVersion 3.14.7-quad.10> - xxxxx/apps/openpype/latest/linux
2023-02-23 16:16:11:  0: PYTHON: *** Latest compatible version found is xxxxx/apps/openpype/latest/linux in <OpenPypeVersion 3.14.7-quad.10>
2023-02-23 16:16:11:  0: PYTHON: --- OpenPype executable: xxxxx/apps/openpype/latest/linux/openpype_console
2023-02-23 16:16:11:  0: PYTHON: >>> Temporary path: /tmp/20230223151611402060_0174f8f2-b38d-11ed-b1c0-b8cef6fc1a5a.json
2023-02-23 16:16:11:  0: PYTHON: >>> Executing: xxxxx/apps/openpype/latest/linux/openpype_console --headless extractenvironments /tmp/20230223151611402060_0174f8f2-b38d-11ed-b1c0-b8cef6fc1a5a.json --project TEST_OP_PUB --asset 001_0010 --task Reference --app nuke/14-0 --envgroup farm
2023-02-23 16:16:31:  0: PYTHON: >>> Loading file ...
2023-02-23 16:16:31:  0: INFO: Setting Process Environment Variable HOSTNAME to nodexxxx
...
2023-02-23 16:16:31:  0: INFO: Setting Process Environment Variable OPENPYPE_MONGO to mongodb://xxxxx
...
2023-02-23 16:16:31:  0: PYTHON: >>> Removing temporary file
2023-02-23 16:16:31:  0: PYTHON: >> Injection end.
2023-02-23 16:16:31:  0: INFO: Scrubbing the LD and DYLD LIBRARY paths
2023-02-23 16:16:31:  0: INFO: Prepping OFX cache
2023-02-23 16:16:31:  0: INFO: Checking Nuke temp path: /var/tmp/nuke-u907700517
2023-02-23 16:16:31:  0: INFO: Path already exists
2023-02-23 16:16:31:  0: INFO: OFX cache prepped
2023-02-23 16:16:31:  0: INFO: Starting monitored managed process Nuke
2023-02-23 16:16:31:  0: INFO: Enable Path Mapping: True
2023-02-23 16:16:31:  0: INFO: Stdout Redirection Enabled: True
...
```

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py (hunk 1 above)
Comment (mre7a) [reply openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:205]: I think so. Maybe the log I added answers the question.

**Comment by iLLiCiTiT** [comment]
Would help if you would call `self.SetProcessEnvironmentVariable("OPENPYPE_MONGO", mongo_url)` and then use `SpawnProcess`?

I'm really worried about using `subprocess.Popen` in Deadline plugins because it caused us a lot of time in past.

**Comment by mre7a** [comment]
> Would help if you would call `self.SetProcessEnvironmentVariable("OPENPYPE_MONGO", mongo_url)` and then use `SpawnProcess`?

It does not work. Same problem. I think it adds the variable to the render process and not the spawned process. 

> I'm really worried about using `subprocess.Popen` in Deadline plugins because it caused us a lot of time in past (and broken production).

I understand, I will try to test other solutions and find out why it doesn't work on Centos. Maybe Deadline settings ?
I will also try to test on Rockey linux.

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py [hunk 4]
```diff
@@ -359,11 +359,11 @@ def inject_openpype_environment(deadlinePlugin):
 
         args_str = subprocess.list2cmdline(args)
         print(">>> Executing: {} {}".format(exe, args_str))
-        process = ProcessUtils.SpawnProcess(
-            exe, args_str, os.path.dirname(exe)
+        process_exitcode = deadlinePlugin.RunProcess(
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:362]: Really, only this?

**Comment by mre7a** [comment]
I tried RunProcess from DeadlinePlugin, and it works fine for us on Centos7. 
And again, I did not test it on Windows.

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py (hunk 4 above)
Comment (mre7a) [reply openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:362]: I verified again, and just changing this solves our problem. I can't see the difference between the two.

Comment (iLLiCiTiT) [reply openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:362]: Ok, we will test on linux and let you know.

**Review by kalisp — APPROVED** [review]
Tested it on Windows, it doesn't seem to break anything.

**Pull request merged**